feat: add `cdk watch` demo content

kaizen3031593 opened the pull request:
Adds a `CDK Watch` page to the workshop demonstrating faster personal deployments for the lambda asset code.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

kaizen3031593 commented:
Still a WIP because I haven't yet added any photos or done this for languages other than TypeScript, but wanted to finalize the structure before I go that far.

It took me about 3 minutes to set up the workshop repo and get it running locally, and that's probably the best experience to see how this looks.

kaizen3031593 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,177 @@
+
+## Faster Personal Deployments
+
+{{% notice info %}} This section is not necessary to complete the workshop, but we
+recommend that you take the time to see how `cdk watch` can speed up your personal
+deployments.
+{{% /notice %}}
```
I still like where we're putting the CDK Watch stuff, but I do think it's a bit disruptive and not necessary for people to complete the workshop, so I wanted to add this disclaimer here.

skinny85 replied:
Sure. I don't mind it 🙂.

kaizen3031593 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,177 @@
+hotswap the lambda asset code, so it will bypass a CloudFormation deployment and
+deploy directly to the Lambda service instead.
+
+How fast did deployment take?
```
Haha.. can some one actually suggest a good way to time an actual deployment. I saw @corymhall use `esbuild`, is that the general way to go?

skinny85 replied:
I actually thought it would be awesome if we printed out, after each deployment in `cdk watch` mode, the total time the deployment took. Like we do now for the `cdk deploy` command, which ends in something like:

```
✨  Done in 78.80s.
```

Would implementing that be something you might be interested in @kaizen3031593? 🙂

kaizen3031593 replied:
So `cdk deploy` doesn't actually do this I don't think. It is something that's probably useful though.

skinny85 replied:
[It definitely does](https://www.youtube.com/watch?v=XBfgvXEaUz0&t=119s) 🙂.

kaizen3031593 replied:
😂 i was literally watching that just now and going crazy. `yarn cdk deploy` spits out the total time afterwards. but `cdk deploy` does not. I confirmed that behavior on my CDK app but can you explain why that's the case?

kaizen3031593 replied:
![Screen Shot 2021-12-29 at 3 16 04 PM](https://user-images.githubusercontent.com/36202692/147700055-0179cc59-e7e0-4f07-a328-da1dde8a26c3.png)

skinny85 replied:
Oh, it's probably a Yarn feature then, and not a feature of the CDK CLI 😃, like I thought it was. Good catch!

Then yeah, let's add it to `cdk deploy` / `cdk watch`. I'll make an issue for it.

skinny85 replied:
https://github.com/aws/aws-cdk/issues/18213

comcalvi commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,179 @@
+development purposes. Never use CDK Watch for your production deployments!
+{{% /notice %}}
+
+We can speed up that deployment time with `cdk watch`, which watches your
+CDK application and looks for hotswappable changes. Which `cdk watch` finds a
+change that can be hotswapped, it bypasses the CloudFormation deployment entirely
+and instead makes the necessary SDK call to update your code for you. Changes to 
+your AWS Lambda asset code, AWS Step Function State Machine definition, and other
+non-resource changes have the potential to be hotswapped. Here, we will try
+`cdk watch` with your AWS Lambda asset code.
```
I would briefly mention the `--hotswap` flag here, and then go into more detail about how it works later on. I would change this section to something like:

"We can speed up that deploy time with `cdk watch`, which monitors your code and assets for changes and attempts to perform a deployment automatically when a change is detected. By default, `cdk watch` will use the `--hotswap` flag, which inspects the changes and determines if those changes can be hotswapped. If they can, the CDK CLI will use AWS service APIs to directly make the changes; otherwise it will fall back to performing a full CloudFormation deployment. To disable this behavior, pass `--no-hotswap` to `cdk watch`."

skinny85 replied:
I like Calvin's suggestion!

comcalvi commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
````diff
@@ -0,0 +1,179 @@
+matter what you change it to, but let's say we want it to say
+`"Good Afternoon, CDK"` now.
+
+Once you click save or hit `CTRL+s`, `cdk watch` will recognize that your file has
+changed and trigger a new deployment. In this case, it will recognize that we can
+hotswap the lambda asset code, so it will bypass a CloudFormation deployment and
+deploy directly to the Lambda service instead.
+
+How fast did deployment take?
+
+```
+INSERT PHOTO HERE
+```
````
I would modify this a bit, something like:

"Once you click save or hit `CTRL+s`, `cdk watch` will recognize that your file has changed and trigger a new deployment. 

```
INSERT PHOTO HERE
```

The warning message indicates that this lambda code change is hotswappable, so it will bypass CloudFormation and go directly to the Lambda Service API. This introduces drift between your CloudFormation template and your deployed application code. Because of this drift, hotswapping should never be used in a production environment. Hotswap deployments are faster, but lack the robust safety features of CloudFormation deployments. To disable hotswapping, pass the `--no-hotswap` flag to `cdk watch`."

skinny85 replied:
I feel like we addressed this already above, with hotswapping. I don't think it makes sense to repeat all of that here.

It's a good idea to mention `cdk watch --no-hotswap` though.

comcalvi reviewed (changes requested):
All in all, looks really good! I have a few small comments. My suggestions may be too verbose for the cdk workshop, and it might be better to just link the users to the blog post. Thoughts on this?

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,179 @@
+
+## CDK Watch
+
+{{% notice info %}} This command deliberately introduces drift in CloudFormation 
+stacks in order to speed up deployments. For this reason, only use it for 
+development purposes. Never use CDK Watch for your production deployments!
+{{% /notice %}}
```
This is actually only a feature of hotswapping, not of `cdk watch`, and you can use "watch" without hotswapping (`cdk watch --no-hotswap`).

We need to make sure the distinction between the two is clearly explained here.

kaizen3031593 replied:
do we need to add more beyond what Calvin has already suggested re: `--hotswap` flag and `--no-hotswap` option?

skinny85 replied:
I think that's enough info there.

(I'll probably change my mind once I see it all written out, but for now it's good enough 😛)

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,179 @@
+This will trigger an initial deployment and immediately begin observing the files
+you've specified in `cdk.json`.
+
+For the purposes of this exercise, make sure that the initial deployment is
+finished before proceeding. Otherwise, the changes you make will be queued for
+deployment and released _after_ the first deployment finishes.
```
Not sure we need this paragraph - even if the changes come during, they will still be picked up, so I don't think it makes sense to add this disclaimer.

skinny85 reviewed (commented):
Looks really good! Some minor comments, and I also agree with Calvin's comments.

kaizen3031593 commented:
A lot of feedback centered around demonstrating `cdk watch` and `hotswap` as separate features which I'll admit I was pretty fuzzy on as well. Because of this I thought it was prudent to add a new section detailing `cdk deploy --hotswap`, which we can time and compare with `cdk deploy` and workshop users can ooh and ahh at how fast it is.

Then I think it makes a lot more sense to show `cdk watch` functionality once they have already seen what `hotswap` is. At any rate, hoping for another review and I'll add in those screenshots tomorrow.

comcalvi commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
````diff
@@ -0,0 +1,213 @@
+INSERT OUTPUT HERE
+```
+
+## CDK Deploy Hotswap
````
I'm not sure I like the title, maybe something like `## Performing a Hotswap Deployment`? `## CDK Deploy --hotswap`? `Hotswap Deployments`?

skinny85 replied:
I think `## Hotswap deployments` is a good title.

comcalvi commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,213 @@
+the changes; otherwise it will fall back to performing a full CloudFormation
+deployment.
+
+Here, we will use `cdk deploy --hotswap` to detect a hotswappable change to your 
```
```suggestion
Here, we will use `cdk deploy --hotswap` to deploy a hotswappable change to your 
```

skinny85 replied:
+1

comcalvi reviewed (changes requested):
Nothing but minor comments from me!

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,213 @@
+weight = 300
++++
+
+## Faster Personal Deployment
```
I don't think we need to uppercase all letters in the title here:

```suggestion
## Faster personal deployments
```

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,213 @@
+We really only need to update our lambda code. It would be great if we had
+some other mechanism for doing only that...
+
+## See how long 'CDK Deploy' takes
```
```suggestion
## Timing `cdk deploy`
```

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,213 @@
+Here, we will use `cdk deploy --hotswap` to detect a hotswappable change to your 
+AWS Lambda asset code.
+
+## See how long 'CDK Deploy --Hotswap' takes
```
```suggestion
## Timing `cdk deploy --hotswap`
```

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,213 @@
+
+Now you're all set to start watching!
+
+## See how long 'CDK Watch' takes
```
```suggestion
## Timing `cdk watch`
```

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,213 @@
+This will trigger an initial deployment and immediately begin observing the files
+you've specified in `cdk.json`.
+
+Let's change our lambda asset code in `lambda/hello.js` one more time. It doesn't
+matter what you change it to, but let's say we want it to say
+`"Good Afternoon, CDK"` now.
```
Let's be more specific here maybe? I feel like just saying "make any change" each time doesn't "jive" well with the rest of the workshop.

How about we say to make a specific change each time, and we use the nice highlighting feature to show clearly in the example what line(s) did we change each time?

kaizen3031593 replied:
the highlighting feature is cool, didn't even realize how we did that.

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,213 @@
+matter what you change it to, but let's say we want it to say
+`"Good Afternoon, CDK"` now.
+
+Once you click save or hit `CTRL+s`, `cdk watch` will recognize that your file has
```
Let's just say "Once you save the changes to your Lambda code file", or something.

skinny85 reviewed (commented):
This is shaping up really nicely! A few minor comments.

kaizen3031593 commented:
Thanks for the helpful feedback up to this point, @comcalvi + @skinny85! I'm ready for another round of review -- resolved all of the comments and added in screenshots and console statements. Say the word and I will reproduce the same file for the other languages as well!

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,298 @@
+can speed up your personal deployments.
+{{% /notice %}}
+
+It's great that we have a working lambda! But what if you want to tweak the lambda
```
Let's us "we" consistently:

```suggestion
It's great that we have a working lambda! But what if we want to tweak the lambda
```

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
````diff
@@ -0,0 +1,298 @@
+```
+
+Wow, deploying a hotswapped change took 3 seconds, while a full deployment took 67 seconds! 
+But take a look and read the warning message thoroughly - it's important!:
````
I think we don't need the colon here, the exclamation mark is good enough:

```suggestion
But take a look and read the warning message thoroughly - it's important!
```

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,298 @@
+
+## Modify your `cdk.json` file
+
+When the `cdk watch` command runs, the files that it observes is determined by the
```
```suggestion
When the `cdk watch` command runs, the files that it observes are determined by the
```

kaizen3031593 replied:
good catch :)

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,298 @@
+Each entry is interpreted as a path relative to the location of the `cdk.json`
+file. Globs, both `*` and `**`, are allowed to be used.
+
+Your `cdk.json` file should look like this:
```
```suggestion
Your `cdk.json` file should look similar to this:
```

skinny85 commented on workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md:
```diff
@@ -0,0 +1,298 @@
+
+3. The code should be loaded onto the screen. Did your change show up?
+
+    ![](./lambda-5.png)
```
You might consider also testing that the function changed by calling it, and observing the result it returned. That's super minor though, this is fine too if you want to keep it as-is.

kaizen3031593 replied:
We do exactly this (test the result) in the previous page where we set up the lambda. So I think it is fine to keep it as-is here; feels like observing the code change is evidence enough.

skinny85 reviewed (commented):
Looks great @kaizen3031593! A few minor comments, but I think we're ready to add this to the other languages.

kaizen3031593 commented:
Added similar content to all other languages as well. Changed the lambda to python to match the python section. Also, other languages did not need any changes to `cdk.json`, so I updated that section for each as well. Can't think of any other special cases (looks good to me).

Discovered that a 'notion info' tag wasn't working in the hello-lambda page so I fixed that as well.

skinny85 reviewed (approved):
Looks awesome, thanks @kaizen3031593!

comcalvi reviewed (approved):
LGTM! Nice work.
